modules/sugar/src/main/ts/ephox/sugar/api/node/ShadowDom.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  Document, Element as DomElement,
+  HTMLElement, HTMLElementTagNameMap,
+  Node as DomNode,
+  ShadowRoot
+} from '@ephox/dom-globals';
+import * as Node from './Node';
+import * as Head from './Head';
+import * as Body from './Body';
+import { Option, Type } from '@ephox/katamari';
+import Element from './Element';
+
+export type RootNode = Element<Document | ShadowRoot>;
+
+export const isShadowRoot = (dos: RootNode): dos is Element<ShadowRoot> =>
+  Node.isDocumentFragment(dos);
+
+export const isDocument = (dos: RootNode): dos is Element<Document> =>
+  Node.isDocument(dos);
+
+/**
+ * Does the browser support shadow DOM?
+ *
+ * NOTE: Node.getRootNode() and Element.attachShadow don't exist on IE11 and pre-Chromium Edge.
+ */
+export const isSupported = (): boolean =>
+  Type.isFunction((DomElement.prototype as any).attachShadow) &&
+  Type.isFunction((DomNode.prototype as any).getRootNode);
+
+export const getRootNode = (e: Element<DomNode>): RootNode => {
+  if (isSupported()) {
+    return Element.fromDom((e.dom() as any).getRootNode());
+  } else {
+    // ownerDocument returns null for a document, and Element.fromDom requires non-null,
+    // so we have to check if the element is a document.
+    return Node.isDocument(e) ? e : Element.fromDom(e.dom().ownerDocument);
+  }
+};
+
+/**
+ * If this is a Document, return it.
+ * If this is a ShadowRoot, return its parent document.
+ */
+export const actualDocument = (dos: RootNode): Element<Document> =>
```

Comment:
`getDocument` instead maybe? at the very least for consistency, this should probably start with `get`.

Reply:
See, a RootNode can be a document or a shadowroot. `getShadowRoot` returns an `Option`, but this one always returns a `Document`. 

So, we probably need 3 functions (not sure of names):
1. `RootNode -> Option<Document>`
2. `RootNode -> Option<ShadowRoot>`
3. `RootNode -> Document`.

Any ideas?

Reply:
For now, 
1. Don't need this function
2. `ShadowRoot.getShadowRoot`
3. `Traverse.documentOrOwner`

Happy to take other suggestions on naming.